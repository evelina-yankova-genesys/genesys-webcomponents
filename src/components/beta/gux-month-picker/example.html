<h1>gux-month-picker</h1>

<div class="container">
  <h2>Default</h2>
  <gux-month-picker-beta value="2022-01"></gux-month-picker-beta>

  <h2>With Label</h2>
  <gux-month-picker-beta label="Select Month"></gux-month-picker-beta>

  <h2>Min & Max</h2>
  <gux-month-picker-beta
    value="2022-01"
    min-date="2021-03"
    max-date="2023-09"
  ></gux-month-picker-beta>

  <h2>Languages</h2>

  <h3>Polish</h3>
  <gux-month-picker-beta lang="pl"></gux-month-picker-beta>

  <h3>Spanish</h3>
  <gux-month-picker-beta lang="es"></gux-month-picker-beta>

  <h3>Korean</h3>
  <gux-month-picker-beta lang="ko"></gux-month-picker-beta>
</div>

<style>
  .container {
    width: 150px;
  }
</style>
