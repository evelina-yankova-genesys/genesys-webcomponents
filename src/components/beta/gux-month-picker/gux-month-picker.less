@import (reference) '../../../style/mixins.less';
@import (reference) '../../../style/spacing.less';
@import (reference) '../../../style/typography.less';

.gux-month-picker-beta {
  &.gux-disabled {
    pointer-events: none;
    cursor: default;

    .gux-month-picker {
      opacity: 0.5;
    }
  }

  .gux-month-picker-label {
    .gux-label();
  }

  .gux-month-picker {
    display: flex;
    padding: @spacing-2xs @spacing-xs @spacing-2xs @spacing-xs;
    line-height: 1.6667;
    cursor: pointer;
    background-color: @gux-grey-90;
    .form-field-border();

    .gux-month-display {
      flex: 1 1 auto;
      align-self: auto;
      order: 0;
      width: 100px;
      padding-right: @spacing-xs;
      font-family: inherit;
      font-size: 12px;
      color: @gux-black-50;
      text-align: left;
    }

    .gux-calendar-toggle-button {
      flex: 0 1 auto;
      align-self: auto;
      order: 0;
      padding: 2px;
      color: @gux-black-90;
      background: transparent;
      border: none;

      &:not(:disabled):focus,
      &:not(:disabled):hover {
        color: @gux-black-50;
        cursor: pointer;
      }

      gux-icon {
        width: 16px;
        height: 16px;
      }
    }
  }

  gux-month-calendar-beta {
    z-index: 1;
    display: none;
  }

  &.gux-active {
    .gux-month-picker {
      .gux-calendar-toggle-button {
        color: @gux-blue-60;

        &:not(:disabled):focus,
        &:not(:disabled):hover {
          color: @gux-blue-60;
        }
      }
    }

    gux-month-calendar-beta {
      display: flex;
    }
  }
}

.gux-sr-only {
  .gux-sr-only();
}
