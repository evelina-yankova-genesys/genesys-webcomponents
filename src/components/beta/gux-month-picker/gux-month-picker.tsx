/* eslint-disable @typescript-eslint/restrict-template-expressions */
import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Watch,
  Prop,
  State
} from '@stencil/core';

import { trackComponent } from '../../../usage-tracking';
import { OnClickOutside } from '../../../utils/decorator/on-click-outside';
import { buildI18nForComponent, GetI18nValue } from '../../../i18n';
import translationResources from './i18n/en.json';
import { getTranslatedMonthsLong } from '../gux-month-calendar/gux-month-calendar.service';

@Component({
  styleUrl: 'gux-month-picker.less',
  tag: 'gux-month-picker-beta',
  shadow: true
})
export class GuxMonthPicker {
  i18n: GetI18nValue;

  @Element()
  private root: HTMLElement;
  monthCalendarElement: HTMLGuxMonthCalendarBetaElement;

  /**
   * The month picker current value
   */
  @Prop({ mutable: true })
  value: string = '';

  /**
   * Indicate if the month picker is disabled or not
   */
  @Prop()
  disabled: boolean = false;

  @Prop()
  label: string;

  /**
   * The min date selectable
   */
  @Prop()
  minDate: string = '';

  /**
   * The max date selectable
   */
  @Prop()
  maxDate: string = '';

  @State()
  active: boolean = false;

  @State()
  yearLabel: number;

  @State()
  monthLabel: string;

  @State()
  translatedMonthsLongArray: Array<string> = [];

  /**
   * Triggered when user selects a month
   */
  @Event()
  input: EventEmitter<string>;

  async componentWillLoad() {
    trackComponent(this.root);
    this.i18n = await buildI18nForComponent(this.root, translationResources);

    if (!this.value) {
      const newDate = new Date();
      const year = newDate.getFullYear();
      const month =
        newDate.getMonth().toString().length < 2
          ? `0${newDate.getMonth() + 1}`
          : newDate.getMonth() + 1;
      this.value = `${year}-${month}`;
    }

    this.translatedMonthsLongArray = getTranslatedMonthsLong(this.i18n);
    this.monthLabel =
      this.translatedMonthsLongArray[parseInt(this.value.split('-')[1])];
    this.yearLabel = parseInt(this.value.split('-')[0]);
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  onClickOutside() {
    this.active = false;
  }

  @Watch('active')
  watchActiveCalendar(active: boolean) {
    if (active) {
      void this.monthCalendarElement.resetCalendarView();
    }
  }

  toggleCalendar() {
    this.active = !this.active;
    if (this.active) {
      // Wait for render before focusing preview date
      setTimeout(() => {
        void this.monthCalendarElement.focusPreviewMonth();
      }, 100);
    }
  }

  onMonthCalendarSelect(inputEvent: Event) {
    const monthCalendar = inputEvent.target as HTMLGuxMonthCalendarBetaElement;
    this.value = monthCalendar.value;
    const value = this.value.split('-');

    this.yearLabel = parseInt(value[0]);
    this.monthLabel = this.translatedMonthsLongArray[parseInt(value[1])];

    inputEvent.stopPropagation(); // Don't let both events bubble.
    this.input.emit(this.value);
    this.active = false;
  }

  renderCalendarToggleButton(): JSX.Element {
    return (
      <button
        class="gux-calendar-toggle-button"
        type="button"
        disabled={this.disabled}
        aria-label={this.i18n('toggleCalendar')}
      >
        <gux-icon decorative icon-name="calendar"></gux-icon>
      </button>
    ) as JSX.Element;
  }

  renderCalendar(): JSX.Element {
    return (
      <gux-month-calendar-beta
        ref={(el: HTMLGuxMonthCalendarBetaElement) =>
          (this.monthCalendarElement = el)
        }
        onInput={(event: CustomEvent) => this.onMonthCalendarSelect(event)}
        value={this.value}
        minDate={this.minDate}
        maxDate={this.maxDate}
      />
    ) as JSX.Element;
  }

  render(): JSX.Element {
    return (
      <div
        class={{
          'gux-month-picker-beta': true,
          'gux-active': this.active,
          'gux-disabled': this.disabled
        }}
      >
        <label class="gux-month-picker-label">
          {this.label}
          <span class="gux-sr-only">
            {this.monthLabel} {this.yearLabel}
          </span>
        </label>
        <div class="gux-month-picker" onClick={() => this.toggleCalendar()}>
          <span
            class="gux-month-display"
            aria-label={`${this.monthLabel} ${this.yearLabel}`}
          >
            {this.monthLabel} {this.yearLabel}
          </span>
          {this.renderCalendarToggleButton()}
        </div>
        {this.renderCalendar()}
      </div>
    ) as JSX.Element;
  }
}
