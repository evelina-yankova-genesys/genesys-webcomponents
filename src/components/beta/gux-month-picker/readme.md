# gux-month-picker



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute  | Description                                     | Type      | Default     |
| ---------- | ---------- | ----------------------------------------------- | --------- | ----------- |
| `disabled` | `disabled` | Indicate if the month picker is disabled or not | `boolean` | `false`     |
| `label`    | `label`    |                                                 | `string`  | `undefined` |
| `maxDate`  | `max-date` | The max date selectable                         | `string`  | `''`        |
| `minDate`  | `min-date` | The min date selectable                         | `string`  | `''`        |
| `value`    | `value`    | The month picker current value                  | `string`  | `''`        |


## Events

| Event   | Description                         | Type                  |
| ------- | ----------------------------------- | --------------------- |
| `input` | Triggered when user selects a month | `CustomEvent<string>` |


## Dependencies

### Depends on

- [gux-icon](../../stable/gux-icon)
- [gux-month-calendar-beta](../gux-month-calendar)

### Graph
```mermaid
graph TD;
  gux-month-picker-beta --> gux-icon
  gux-month-picker-beta --> gux-month-calendar-beta
  gux-month-calendar-beta --> gux-icon
  style gux-month-picker-beta fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
