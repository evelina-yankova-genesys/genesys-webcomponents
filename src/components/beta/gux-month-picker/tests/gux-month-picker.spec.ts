import { newSpecPage } from '@stencil/core/testing';

import { GuxMonthPicker } from '../gux-month-picker';

const components = [GuxMonthPicker];
const language = 'en';

describe('gux-month-picker-beta', () => {
  it('should build', async () => {
    const page = await newSpecPage({
      components,
      html: `<gux-month-picker-beta></gux-month-picker-beta>`,
      language
    });
    expect(page.rootInstance).toBeInstanceOf(GuxMonthPicker);
  });

  describe('#render', () => {
    [`<gux-month-picker-beta value="2022-01"></gux-month-picker-beta>`].forEach(
      (input, index) => {
        it(`should render component as expected (${index + 1})`, async () => {
          const page = await newSpecPage({
            components,
            html: input,
            language
          });

          expect(page.root).toMatchSnapshot();
        });
      }
    );
  });
});
