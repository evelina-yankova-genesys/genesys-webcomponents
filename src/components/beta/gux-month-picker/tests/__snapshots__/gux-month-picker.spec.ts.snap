// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-month-picker-beta #render should render component as expected (1) 1`] = `
<gux-month-picker-beta value="2022-01">
  <mock:shadow-root>
    <div class="gux-month-picker-beta">
      <label class="gux-month-picker-label">
        <span class="gux-sr-only">
          January 2022
        </span>
      </label>
      <div class="gux-month-picker">
        <span aria-label="January 2022" class="gux-month-display">
          January 2022
        </span>
        <button aria-label="Toggle month calendar view" class="gux-calendar-toggle-button" type="button">
          <gux-icon decorative="" icon-name="calendar"></gux-icon>
        </button>
      </div>
      <gux-month-calendar-beta maxdate="" mindate="" value="2022-01"></gux-month-calendar-beta>
    </div>
  </mock:shadow-root>
</gux-month-picker-beta>
`;
